acra-censor/common/common.go

```diff
@@ -150,3 +176,145 @@ func ParsePatterns(rawPatterns []string) ([]sqlparser.Statement, error) {
 	}
 	return outputPatterns, nil
 }
+
+// TrimStringToN trims query to N chars.
+func TrimStringToN(query string, n int) string {
+	if len(query) <= n {
+		return query
+	}
+	return query[:n]
+}
+
+// HandleRawSQLQuery returns a normalized (lowercases SQL commands) SQL string,
+// and redacted SQL string with the params stripped out for display.
+// Taken from sqlparser package
+func HandleRawSQLQuery(sql string) (normalizedQuery, redactedQuery string, parsedQuery sqlparser.Statement, err error) {
+	bv := map[string]*querypb.BindVariable{}
+	sqlStripped, _ := sqlparser.SplitMarginComments(sql)
+
+	// sometimes queries might have ; at the end, that should be stripped
+	sqlStripped = strings.TrimSuffix(sqlStripped, ";")
+
+	stmt, err := sqlparser.Parse(sqlStripped)
+	outputStmt, _ := sqlparser.Parse(sqlStripped)
+	if err != nil {
+		return "", "", nil, ErrQuerySyntaxError
+	}
+
+	normalizedQ := sqlparser.String(stmt)
+
+	// redact and mask VALUES
+	sqlparser.Normalize(stmt, bv, ValueMask)
+	redactedQ := sqlparser.String(stmt)
+
+	return normalizedQ, redactedQ, outputStmt, nil
+}
+
+// CheckPatternsMatching evaluates if parsed query matches specified set of patterns
+func CheckPatternsMatching(patterns []sqlparser.Statement, parsedQuery sqlparser.Statement) bool {
+	for _, pattern := range patterns {
+		if checkSinglePatternMatch(parsedQuery, pattern) {
+			return true
+		}
+	}
+	return false
+}
+
+// CheckExactQueriesMatch evaluates if query presents in set of queries
+func CheckExactQueriesMatch(normalizedQuery string, setOfQueries map[string]bool) bool {
+	if !setOfQueries[normalizedQuery] {
+		return false
+	}
+	return true
+}
+
+// CheckTableNamesMatch evaluates if query contains table presented in specified set of tables
+func CheckTableNamesMatch(parsedQuery sqlparser.Statement, setOfTables map[string]bool) (bool, bool) {
+	atLeastOneTableNameMatch := false
+	allTableNamesMatch := false
+
+	switch parsedQuery.(type) {
+	case *sqlparser.Select:
+		selectQuery := parsedQuery.(*sqlparser.Select)
+		atLeastOneTableNameMatch, allTableNamesMatch = checkTableExprsMatch(selectQuery.From, setOfTables)
+		break
+	case *sqlparser.Insert:
+		insertQuery := parsedQuery.(*sqlparser.Insert)
+		if setOfTables[insertQuery.Table.Name.String()] {
+			atLeastOneTableNameMatch = true
+			allTableNamesMatch = true
+		} else {
+			atLeastOneTableNameMatch = false
+			allTableNamesMatch = false
+		}
+		break
+	default:
+		//TODO other query types
+		return false, false
+	}
+
+	return atLeastOneTableNameMatch, allTableNamesMatch
+}
+
+// Tables matchers
+func checkTableExprsMatch(tables sqlparser.TableExprs, setOfTables map[string]bool) (bool, bool) {
+	oneTableMatch := false
+	allTablesMatch := false
+	counter := 0
+	for _, tableExpr := range tables {
+		oneTableMatchInternal, allTablesMatchInternal := checkTableExprMatch(tableExpr, setOfTables)
+		if oneTableMatchInternal {
+			oneTableMatch = true
+			if allTablesMatchInternal {
+				counter++
+			}
```

Comment:
```
else { 
  break
}
```
we don't need to continue if we have one false for `alltablesMatchInternal`

Reply:
To use it both whitelist and blacklist we should have information if AT LEAST ONE table is in list (blacklist blocks table if it's true, allows - otherwise) and if ALL tables are in list (whitelist allows table if it's true and blocks otherwise)

Reply:
yes, but if we have `oneTableMatch == true` and meet first `allTablesMatchInternal == false` then we can stop iteration and return `true, false` because all ``allTablesMatchInternal` will not true anyway

Reply:
Agree